validation.js

```diff
@@ -0,0 +1,281 @@
+// SPDX-FileCopyrightText: 2022 Andre 'Staltz' Medeiros
+//
+// SPDX-License-Identifier: LGPL-3.0-only
+
+const Ref = require('ssb-ref')
+const SSBURI = require('ssb-uri2')
+const classic = require('ssb-classic/format')
+const ssbKeys = require('ssb-keys')
+const isCanonicalBase64 = require('is-canonical-base64')
+const getMsgId = require('./get-msg-id')
+
+const isSignatureRegex = isCanonicalBase64('', '\\.sig.\\w+')
+
+function validateShape(nativeMsg) {
+  if (Array.isArray(nativeMsg)) {
+    if (nativeMsg.length !== 2) {
+      return new Error('invalid message: tuple must have 2 elements')
+    }
+    const [msgVal] = nativeMsg
+    return validateShape(msgVal)
+  }
+
+  if (!nativeMsg || typeof nativeMsg !== 'object') {
+    return new Error('invalid message: not an object')
+  }
+  const msgVal = nativeMsg
+
+  if (typeof msgVal.author === 'undefined') {
+    return new Error('invalid message: must have author')
+  }
+  if (typeof msgVal.previous === 'undefined') {
+    return new Error('invalid message: must have previous')
+  }
+  if (typeof msgVal.sequence === 'undefined') {
+    return new Error('invalid message: must have sequence')
+  }
+  if (typeof msgVal.timestamp === 'undefined') {
+    return new Error('invalid message: must have timestamp')
+  }
+  if (typeof msgVal.hash === 'undefined') {
+    return new Error('invalid message: must have hash')
+  }
+  if (typeof msgVal.content === 'undefined') {
+    return new Error('invalid message: must have content')
+  }
+  if (typeof msgVal.signature === 'undefined') {
+    return new Error('invalid message: must have signature')
+  }
+}
+
+function validateAuthor(msgVal) {
+  if (!SSBURI.isIndexedV1FeedSSBURI(msgVal.author)) {
+    return new Error('invalid message: must have author as an SSB URI')
+  }
+}
+
+function validateSignature(msgVal, hmacKey) {
+  const { signature } = msgVal
+  if (typeof signature !== 'string') {
+    return new Error('invalid message: must have signature as a string')
+  }
+  if (!signature.endsWith('.sig.ed25519')) {
+    // prettier-ignore
+    return new Error('invalid message: signature must end with .sig.ed25519')
+  }
+  if (!isSignatureRegex.test(signature)) {
+    // prettier-ignore
+    return new Error('invalid message: signature must be a canonical base64 string')
+  }
+  if (signature.length !== 100) {
+    // prettier-ignore
+    return new Error('invalid message: signature must be 64 bytes, on feed: ' + msgVal.author);
+  }
+
+  const { data } = SSBURI.decompose(msgVal.author)
+  const keys = { public: data, curve: 'ed25519' }
+  if (!ssbKeys.verifyObj(keys, hmacKey, msgVal)) {
+    // prettier-ignore
+    return new Error('invalid message: signature does not match, on feed: ' + msgVal.author);
+  }
+}
+
+function validateOrder(msgVal) {
+  const keys = Object.keys(msgVal)
```

Comment:
some of these things we might simplify, they were only there because historically we did not have this check so some "bad" classic messages slipped through.